packages/opentelemetry-types/src/trace/span.ts

```diff
@@ -0,0 +1,106 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { SpanContext } from './span_context';
+import { Status } from './status';
+
+/**
+ * An interface that represents a span. A span represents a single operation
+ * within a trace. Examples of span might include remote procedure calls or a
+ * in-process function calls to sub-components. A Trace has a single, top-level
+ * "root" Span that in turn may have zero or more child Spans, which in turn
+ * may have children.
+ */
+export interface Span {
+  /**
+   * Returns the SpanContext object associated with this Span.
+   *
+   * @returns the SpanContext object associated with this Span.
+   */
+  context(): SpanContext;
+
+  // /**
+  //  * # TODO
+  //  * Returns the Tracer object used to create this Span.
+  //  * https://github.com/open-telemetry/opentelemetry-specification/issues/21
+  //  */
+  // tracer(): Tracer;
+
+  /**
+   * Sets an attribute to the span.
+   *
+   * @param key the key for this attribute.
+   * @param value the value for this attribute. If the value is a typeof object
+   *     it has to be JSON.stringify-able otherwise it will raise an exception.
+   */
+  setAttribute(key: string, value: string | number | boolean | object): void;
+
+  /**
+   * Adds an event to the Span.
+   *
+   * @param name the name of the event.
+   * @param [attributes] the attributes that will be added; these are
+   *     associated with this event.
+   */
+  addEvent(
+    name: string,
+    attributes?: { [key: string]: string | number | boolean | object }
+  ): void;
+
+  /**
+   * Adds a link to the Span.
+   *
+   * @param spanContext the context of the linked span.
+   * @param [attributes] the attributes that will be added; these are
+   *     associated with this link.
+   */
+  addLink(
+    spanContext: SpanContext,
+    attributes?: { [key: string]: string | number | boolean | object }
+  ): void;
+
+  /**
+   * Sets a status to the span. If used, this will override the default Span
+   * status. Default is 'OK'.
+   *
+   * @param status the Status to set.
+   */
+  setStatus(status: Status): void;
+
+  /**
+   * Updates the Span name.
+   *
+   * @param name the Span name.
+   */
+  updateName(name: string): void;
+
+  /**
+   * Marks the end of Span execution and sets the current time as the span's
+   * end time.
+   *
+   * Call to End of a Span MUST not have any effects on child spans. Those may
+   * still be running and can be ended later.
+   */
+  end(): void;
```

Comment:
This should accept a timestamp since it's the only way to support some frameworks (i.e. Apollo Tracing).

Reply:
Makes sense, done.

Reply:
Shouldn't we change the spec for this?
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/tracing-api.md#end
> There MUST be no parameter.

Reply:
I would say yes as the same issue applies to every language. The same will be true also for the ability to set the start time.

Reply:
https://github.com/open-telemetry/opentelemetry-specification/pull/120